### PR TITLE
AM2315C Temperature + Humidity Sensor

### PR DESCRIPTION
**Pull request by swoboda1337**
# What does this implement/fix?

Adds support for the am2315c temp and humidity sensor. This sensor is conveniently packaged for use in HVAC (ie duct mounting). Comes with duct mounting hardware. Reasonably priced. I have used it for a few months now and think it would be useful to others. 

https://www.adafruit.com/product/5182

[D][sensor:094]: 'Duct Temperature': Sending state 20.52689 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20432 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.51163 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20966 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.50819 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20937 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.50266 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20794 % with 1 decimals of accuracy

Code was based off this project but has been cleaned up and ported to esphome https://github.com/RobTillaart/AM2315C

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3625

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: am2315c
    temperature:
      name: "Temperature"
    humidity:
      name: "Humidity"
    update_interval: 60s

i2c:
  sda: 21
  scl: 22
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by swoboda1337** [comment]
[D][sensor:094]: 'Duct Temperature': Sending state 20.52689 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20432 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.51163 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20966 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.50819 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20937 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.50266 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.20794 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.48931 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.23359 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.46509 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.23016 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.44144 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.26421 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.43247 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.25257 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.42885 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.27336 % with 1 decimals of accuracy
[D][sensor:094]: 'Duct Temperature': Sending state 20.41054 °C with 1 decimals of accuracy
[D][sensor:094]: 'Duct Humidity': Sending state 35.29406 % with 1 decimals of accuracy

File: esphome/components/am2315c/am2315c.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+class AM2315C : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void dump_config() override;
+  void update() override;
+  void setup() override;
+  float get_setup_priority() const override;
+
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { humidity_sensor_ = humidity_sensor; }
```
Comment (jesserockz) [inline esphome/components/am2315c/am2315c.h]:
```suggestion
  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { this->temperature_sensor_ = temperature_sensor; }
  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { this->humidity_sensor_ = humidity_sensor; }
```

File: esphome/components/am2315c/am2315c.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    if (!this->reset_register_(0x1C)) {
+      this->mark_failed();
+      return;
+    }
+    if (!this->reset_register_(0x1E)) {
+      this->mark_failed();
+      return;
+    }
+    delay(10);
```
Comment (jesserockz) [inline esphome/components/am2315c/am2315c.cpp]: This delay is not needed
```suggestion
```

File: esphome/components/am2315c/am2315c.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    uint8_t data[7];
+    if (this->read(data, 7) != i2c::ERROR_OK) {
+      ESP_LOGE(TAG, "Read failed!");
+      this->mark_failed();
+      return;
+    }
+
+    // check for all zeros
+    bool zeros = true;
+    for (unsigned char i : data) {
```
Comment (jesserockz) [inline esphome/components/am2315c/am2315c.cpp]:
```suggestion
    for (uint8_t i : data) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**